File: 3_happy_ladybugs/py/ladybugs.py
import sys


def wrap(board):
    return "YES" if ladybug(board) else "NO"


def ladybug(board):
    if "_" in board:
        counts = {}
        for char in board:
            if char != "_":
                if char not in counts:
                    counts[char] = 0
                counts[char] += 1
        return all([count >= 2 for count in counts.values()])
    else:  # no moves available, all bugs must be happy at start or never at all
        all_happy = len(board) > 1 and board[0] == board[1] and board[-1] == board[-2]
        for idx, bug in enumerate(board[1:-1]):
            idx += 1  # we started at 1
            all_happy = all_happy and (bug == board[idx - 1] or bug == board[idx + 1])
            if not all_happy:
                break
        return all_happy


Q = int(input().strip())
for a0 in range(Q):
    n = int(input().strip())
    b = input().strip()
    print(wrap(b))

File: 3_happy_ladybugs/go/ladybugs.go
package main

import (
	"fmt"
	"strings"
)

func ladybug(board string) bool {
	if strings.Contains(board, "_") {
		// build count map
		counts := map[uint8]int{};
		for idx := 0; idx < len(board); idx += 1 {
			ch := board[idx]
			if ch != '_' {
				_, charExists := counts[ch]
				if !charExists {
					counts[ch] = 0
				}
				counts[ch] += 1
			}
		}
		// check count map
		success := true
		for _, value := range counts {
			if value < 2 {
				success = false
			}
		}
		return success
	} else {
		// board must start happy else it is sad
		if len(board) < 2 {
			return false
		}
		lastIdx := len(board) - 1
		allHappy := board[0] == board[1] && board[lastIdx] == board[lastIdx-1]
		for idx := 1; idx < len(board) - 1; idx += 1 {
			allHappy = allHappy && (board[idx] == board[idx + 1] || board[idx] == board[idx-1])
		}
		return allHappy
	}
}

func wrap(board string) string {
	if ladybug(board) {
		return "YES"
	} else {
		return "NO"
	}
}

func main() {
	var numGames int;
	fmt.Scan(&numGames)
	for idx := 0; idx < numGames; idx++ {
		var boardLen int;
		var board string;
		fmt.Scan(&boardLen)
		fmt.Scan(&board)
		fmt.Println(wrap(board))
	}
}